src/lib/markbind/src/preprocessors/variablePreprocessor.js

```diff
@@ -0,0 +1,468 @@
+const cheerio = require('cheerio');
+const fs = require('fs');
+const path = require('path');
+
+const _ = {};
+_.clone = require('lodash/clone');
+_.cloneDeep = require('lodash/cloneDeep');
+_.has = require('lodash/has');
+_.isArray = require('lodash/isArray');
+_.isEmpty = require('lodash/isEmpty');
+
+const md = require('../lib/markdown-it');
+const utils = require('../utils');
+const njUtil = require('../utils/nunjuckUtils');
+const urlUtils = require('../utils/urls');
+
+const {
+  ATTRIB_CWF,
+  IMPORTED_VARIABLE_PREFIX,
+} = require('../constants');
+
+
+/**
+ * Variable preprocessor to recursively extract and render pages' variables.
+ * Houses a site-wide variable cache.
+ */
+class VariablePreprocessor {
```

Comment:
centerpoint of variable processing; There's generally 4 categories of methods here ( demarcated with dividers below )
1. site variable storage (get/set) type of methods
2. site variable-only rendering type of methods ( using (1) )
3. page variable storage ( the 3 static variables originally from `Parser` - this is likely still to be revamped in part 2 ) - unlike (1), this also includes 'extraction type' methods ( methods that to sift out the outer page variables and inner page variables ) - here, not much is changed for now ( will annotate where important ), but will be in part 2
4. combined rendering methods (new) - the main benefit of this part 1 pr, along with (2) - which allows reducing boilerplate (main pr post for example). Still a work in progress, with part 2 aiming to reduce this even further, but much less than (3).